exporter/googleservicecontrolexporter/exporter.go

```diff
@@ -0,0 +1,499 @@
+// [2] https://cloud.google.com/service-infrastructure/docs/service-control/reference/rest/v1/Operation
+func (e *Exporter) createReportRequest(rms pmetric.ResourceMetricsSlice) *scpb.ReportRequest {
+	now := time.Now()
+	request := scpb.ReportRequest{
+		Operations:      make([]*scpb.Operation, 0),
+		ServiceConfigId: e.serviceConfigID,
+		ServiceName:     e.serviceName,
+	}
+
+	for i := 0; i < rms.Len(); i++ {
+		rm := rms.At(i)
+		resource := rm.Resource()
+		scopeMetricsSlice := rm.ScopeMetrics()
+		resourceAttributes := attributesToStringMap(resource.Attributes())
+
+		// By default, use consumerID from the exporter configuration.
+		consumerID := e.consumerID
+
+		// Allow users to override the consumerID by providing a resource attribute.
+		if v, found := resourceAttributes[dynamicConsumerAttribute]; found {
+			// Delete the attribute: it is only for Metrics Agent to understand the correct consumer id.
+			// Service Control does not know about this label, and will complain if we send it.
+			delete(resourceAttributes, dynamicConsumerAttribute)
+			consumerID = v
+		}
+
+		for j := 0; j < scopeMetricsSlice.Len(); j++ {
+			metrics := scopeMetricsSlice.At(j).Metrics()
+			for k := 0; k < metrics.Len(); k++ {
+				request.Operations = append(request.Operations, e.createOperation(resourceAttributes, metrics.At(k), now, consumerID))
+			}
+		}
+	}
+
+	return &request
+}
+
+// We create a dedicated Operation for each metric, see go/slm-monitoring-opentelemetry-batching for details.
+func (e *Exporter) createOperation(resourceAttributes map[string]string, metric pmetric.Metric, now time.Time, consumerID string) *scpb.Operation {
+	start := now
+	op := scpb.Operation{
+		ConsumerId:    consumerID,
+		OperationName: operationName,
+		EndTime:       timestamppb.New(now),
+		// These labels are monitored resource labels.
+		// Metric labels are stored in MetricValue proto.
+		Labels:          resourceAttributes,
+		MetricValueSets: make([]*scpb.MetricValueSet, 1),
+		OperationId:     uuid.New(),
+	}
+
+	mvs, st := e.createMetricValueSet(metric)
+	op.MetricValueSets[0] = mvs
+	if !st.IsZero() && st.Before(start) {
+		start = st
+	}
+	op.StartTime = timestamppb.New(start)
+
+	return &op
+}
+
+func (e *Exporter) createMetricValueSet(metric pmetric.Metric) (*scpb.MetricValueSet, time.Time) {
+	vs := &scpb.MetricValueSet{
+		MetricName: metric.Name(),
+	}
+
+	var startTime time.Time
+	var mv []*scpb.MetricValue
+	t := metric.Type()
+	switch t {
+	case pmetric.MetricTypeGauge:
+		mv, startTime = e.createNumericMetricValues(metric.Gauge().DataPoints(), pmetric.AggregationTemporalityUnspecified)
+	case pmetric.MetricTypeSum:
+		mv, startTime = e.createNumericMetricValues(metric.Sum().DataPoints(), metric.Sum().AggregationTemporality())
+	case pmetric.MetricTypeHistogram:
+		mv, startTime = e.createHistogramMetricValues(metric.Histogram())
+	// We don't handle ExponentialHistogram and Summary types, because:
```

Comment:
Lets create an issue for this and track it outside this comment (github issue or an internal one are both fine by me)